Misc xtensa fixups

andyross opened the pull request:
Fixes for issues discovered on ESP-32/S1000 after the merge of the new xtensa timer driver.

Note this also includes a patch cribbed from the x86_64 PR which enables a statically-linked printk handler for very early debugging (the crash on ESP-32 was actually due to an interrupt dispatch goof, and the timer initialization happens well before the UART arrives for printk).

Commit:
misc/printk: Make the default char_out routine weak

Architecture init code and/or HAL layers often have a working
"putchar" routine available long before the Zephyr driver layer is
initialized.

Make the default printk() output a weak symbol, so it can be
overridden on these platforms.

Also remove the kconfig depedency on CONSOLE_HAS_DRIVER, as this is a
non-driver mechanism.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
esp32: Enable an early-boot printk hook

The firmware starts us with an already-initialized UART and a simple
ROM hook.  Use that as the default printk handler so we can log in
very early boot.  It will be overriden later (as it happens, by
another handler based on the same firmware hooks).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Don't call ISRs for disabled interrupts

Xtensa interrupts are handled generically, by testing a set of flagged
interrupts in the INTERRUPT register.  It's not possible to know
exactly which device "caused" an interrupt.

The entry code was dispatching correctly, but it was failing to test
the enable state in INTENABLE.  Such an interrupt will never "fire",
but it might still be flagged, and if we happen to end up handling an
interrupt of the same priority (due to some other device) the entry
handler would incorrectly invoke the disabled interrupt.

Found by dumb luck and a comedy of errors: the recent timer driver
change swapped the counter in use, which changed the interrupt number
to one shared with the I2C driver, whose early interrupts (odd that
this device is interrupting on boot when not in use, but whatever)
would then discover the OTHER timer counter had been flagged and try
to invoke an ISR for that other counter, which was the _irq_spurious()
spurious interrupt handler.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa legacy: Fix stale use of removed kconfigs

The old xtensa layer had an unused/untested facility where it would
apparently try to slave a timer tick to an arbitrary interrupt.  The
legacy headers were still checking the kconfigs used to enable that
even though nothing wants it and the new driver has removed them,
breaking builds on platforms like S1000 that still use the older
layer.

Don't try to finess this as these files are going away.  Just make
them local preprocessor symbols and set them to the default values
they always had.

(Note: the feature doesn't sound like it would have been so bad,
actually.  We should probably crib that idea of having an
"external_tick" driver, but there's no reason for it to have been
arch-specific.)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented on arch/xtensa/include/xtensa_timer.h:
```diff
@@ -31,7 +31,10 @@
+#define USE_INTERNAL_TIMER 1
+#define EXTERNAL_TIMER_IRQ -1
+
+#if USE_INTERNAL_TIMER || (EXTERNAL_TIMER_IRQ < 0)
```
Thought about it but didn't want to break anything.   This file is part of the old assembly layer and isn't included by qemu_xtensa or esp32.  And S1000 just barely missed having been ported due to @dcpleung 's vacation schedule

nashif commented:
```
ccache /opt/xtensa/XtDevTools/install/tools/RF-2016.4-linux/XtensaTools/bin/xt-xcc    CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr_prebuilt.elf -T linker.cmd -Wl,-Map=/home/nashif/Work/github/zephyr/build/zephyr/zephyr.map -u_OffsetAbsSyms -u_ConfigAbsSyms -Wl,--whole-archive ../libapp.a libzephyr.a arch/xtensa/core/startup/libarch__xtensa__core__startup.a lib/libc/minimal/liblib__libc__minimal.a soc/xtensa/intel_s1000/libsoc__xtensa__intel_s1000.a subsys/usb/class/hid/libsubsys__usb__class__hid.a drivers/serial/libdrivers__serial.a -Wl,--no-whole-archive kernel/libkernel.a CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj -L/home/nashif/Work/github/zephyr/build/zephyr -lgcc -lhal -nostdlib -static -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none
libzephyr.a(xt_zephyr.S.obj):(.literal+0x8): undefined reference to `_timer_int_handler'
make[2]: *** [zephyr/CMakeFiles/zephyr_prebuilt.dir/build.make:98: zephyr/zephyr_prebuilt.elf] Error 2
make[2]: Leaving directory '/home/nashif/Work/github/zephyr/build'
make[1]: *** [CMakeFiles/Makefile2:426: zephyr/CMakeFiles/zephyr_prebuilt.dir/all] Error 2
make[1]: Leaving directory '/home/nashif/Work/github/zephyr/build'
make: *** [Makefile:87: all] Error 2
make: Leaving directory '/home/nashif/Work/github/zephyr/build'
```

Commit:
drivers/timer/xtensa_sys_timer: Add hook for old-style interrupt handling

We still have one platform using (for now) the pre-asm2 integration
where the timer interrupt was handled via custom assembly.  It calls a
function named "_timer_int_handler" always, not the one we register
with IRQ_CONNECT.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

nashif commented:
fixes build issues and breakage on esp32.
